[WIP] Fixes #2943 - implements one xDS connection per resource type and config server

dmitri-d opened the pull request:
* rebased
* GrpcMuxFactory.getOrCreateMux uses type_url to differentiate between resources

@htuch: https://github.com/dmitri-d/envoy/commit/aefebec5c2154ef474d76403bf8d8e80c1aa3e45#diff-c0c965f7fd55bbb5ef55f13321e43270R344 is this what you had in mind when you commented on insufficiency of cluster_name/target_uri for map keys?

htuch commented on include/envoy/upstream/grpc_mux_factory.h:
```diff
@@ -0,0 +1,28 @@
+public:
+    virtual ~GrpcMuxFactory() {}
+
+    virtual Config::GrpcMux*
```
Can you add Doxygen comments, thanks.

htuch commented on source/common/upstream/cluster_manager_impl.h:
```diff
@@ -327,6 +328,42 @@ class ClusterManagerImpl : public ClusterManager, Logger::Loggable<Logger::Id::u
+      std::string mux_key; // config_source_name + type_url
+      switch(config_source.grpc_services(0).target_specifier_case()) {
+        case envoy::api::v2::core::GrpcService::kEnvoyGrpc: {
+          mux_key = config_source.grpc_services(0).envoy_grpc().cluster_name() + type_url;
```
I wonder if we should literally require proto equality here. The reasoning here is that there are configuration options, e.g. timeouts, initial metadata, which should ideally be consistent.

dmitri-d commented:
* GrpcMux now starts a new connection when a new subscription is created
* Added Doxygen comments in include/envoy/upstream/grpc_mux_factory.h
* GrpcService_EnvoyGrpc/GrpcService_GoogleGrpc is now used as keys for GrpcMuxes in GrpcMuxFactoryImpl

htuch commented on include/envoy/upstream/grpc_mux_factory.h:
```diff
@@ -0,0 +1,39 @@
+    * @param local_info LocalInfo::LocalInfo local info.
+    * @param async_client Grpc::AsyncClientPtr async client to use for the grpc connection
+    * @param dispatcher event dispatcher.
+    * @param _method fully qualified name of v2 gRPC API bidi streaming method (as per protobuf
```
Nit: `service_method`

htuch commented on include/envoy/upstream/grpc_mux_factory.h:
```diff
@@ -0,0 +1,39 @@
+    virtual ~GrpcMuxFactory() {}
+
+    /**
+    * Returns an existing or creates a new GrpcMux.
```
Can you explain the criteria by which we get existing vs. new?

htuch commented on source/common/upstream/cluster_manager_impl.h:
```diff
@@ -327,6 +329,43 @@ class ClusterManagerImpl : public ClusterManager, Logger::Loggable<Logger::Id::u
+                                    Stats::Scope& scope, const std::string type_url) override {
+
+      uint64_t proto_hash; // config_source_name + type_url
+      switch(config_source.grpc_services(0).target_specifier_case()) {
```
I would just write `proto_hash = MessageUtil::hash(config_source.grpc_services(0)`

htuch commented on source/common/upstream/cluster_manager_impl.h:
```diff
@@ -327,6 +329,43 @@ class ClusterManagerImpl : public ClusterManager, Logger::Loggable<Logger::Id::u
+          NOT_REACHED_GCOVR_EXCL_LINE;
+      }
+
+      const uint64_t mux_key = proto_hash + XXH64(type_url.c_str(), type_url.length() + 1, 0);
```
Maybe `^` instead of `+`, or even better, use proto_hash as the seed.

htuch commented on source/common/upstream/cluster_manager_impl.h:
```diff
@@ -327,6 +329,43 @@ class ClusterManagerImpl : public ClusterManager, Logger::Loggable<Logger::Id::u
+    }
+
+  private:
+    std::unordered_map<uint64_t, Config::GrpcMux*> muxes_;
```
`GrpcMux` should be a `unique_ptr` here to get cleanup via RAII.

dmitri-d commented:
* Updated Doxygen docs for GrpcMuxFactory
* GrpcMuxFactoryImpl now uses unique_ptr<GrpMux> internally
* Modified hash key generation to use GrpcService and use that as a seed in a call to XXH64.
